File: libs/core/src/binder/for_loop_binder.cpp
/**
 * @file for_loop_binder.cpp
 * @author Francesco Corso
 * @date 8/27/2026
 * @brief
 */
module prism.core:binder.for_loop_binder.impl;

import :binder.for_loop_binder;

namespace prism
{

    std::span<Ref<const VariableSymbol>> ForLoopBinder::build_local_variables_impl() const
    {
        if (!syntax_.declaration().has_value())
            return {};

        std::array<Ref<const StatementSyntax>, 1> variables = {*syntax_.declaration()};
        return build_local_variables(variables, *this);
    }
} // namespace prism

File: src/Prism.Core/Binding/ForLoopBinder.cs
using System.Collections.Immutable;
using Prism.Core.Symbols;
using Prism.Core.Syntax;

namespace Prism.Core.Binding;

internal sealed class ForLoopBinder(Binder next, ForStatementSyntax syntax, LabelSymbol label)
    : LoopBinder(next, label)
{
    protected override ImmutableArray<VariableSymbol> BuildLocalVariables()
    {
        return syntax.Declaration is not null
            ? BuildLocalVariables([syntax.Declaration], this)
            : [];
    }
}
